Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp

```diff
@@ -0,0 +1,312 @@
+ *
+ * @param indexingType An index type enum
+ */
+void ScanningWorkspaceBuilder::setIndexingType(
+    const IndexingType indexingType) {
+  if (m_indexingType != IndexingType::Default)
+    throw std::logic_error("Indexing type has been set already.");
+
+  m_indexingType = indexingType;
+}
+
+/**
+ * Verify everything has been set that is required and return the workspace.
+ *
+ * @return Workspace2D with the scanning information set
+ */
+MatrixWorkspace_sptr ScanningWorkspaceBuilder::buildWorkspace() const {
+  validateInputs();
+
+  auto outputWorkspace = create<Workspace2D>(
+      m_instrument, m_nDetectors * m_nTimeIndexes, m_histogram);
+
+  auto &outputDetectorInfo = outputWorkspace->mutableDetectorInfo();
+  outputDetectorInfo.setScanInterval(0, m_timeRanges[0]);
+
+  buildOutputDetectorInfo(outputDetectorInfo);
+
+  if (!m_positions.empty())
+    buildPositions(outputDetectorInfo);
+
+  if (!m_rotations.empty())
+    buildRotations(outputDetectorInfo);
+
+  if (!m_instrumentAngles.empty())
+    buildRelativeRotationsForScans(outputDetectorInfo);
+
+  switch (m_indexingType) {
+  case IndexingType::Default:
+    outputWorkspace->setIndexInfo(
+        Indexing::IndexInfo(m_nDetectors * m_nTimeIndexes));
+    break;
+  case IndexingType::TimeOriented:
+    createTimeOrientedIndexInfo(*outputWorkspace);
+    break;
+  case IndexingType::DetectorOriented:
+    createDetectorOrientedIndexInfo(*outputWorkspace);
+    break;
+  }
+
+  return boost::shared_ptr<MatrixWorkspace>(std::move(outputWorkspace));
+}
+
+void ScanningWorkspaceBuilder::buildOutputDetectorInfo(
+    DetectorInfo &outputDetectorInfo) const {
+  auto mergeWorkspace =
+      create<Workspace2D>(m_instrument, m_nDetectors, m_histogram.binEdges());
+  for (size_t i = 1; i < m_nTimeIndexes; ++i) {
+    auto &mergeDetectorInfo = mergeWorkspace->mutableDetectorInfo();
+    for (size_t j = 0; j < m_nDetectors; ++j) {
+      mergeDetectorInfo.setScanInterval(j, m_timeRanges[i]);
+    }
+    outputDetectorInfo.merge(mergeDetectorInfo);
+  }
+}
+
+void ScanningWorkspaceBuilder::buildRotations(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < m_nDetectors; ++i) {
+    for (size_t j = 0; j < m_nTimeIndexes; ++j) {
+      outputDetectorInfo.setRotation({i, j}, m_rotations[i][j]);
+    }
+  }
+}
+
+void ScanningWorkspaceBuilder::buildPositions(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < m_nDetectors; ++i) {
+    for (size_t j = 0; j < m_nTimeIndexes; ++j) {
+      outputDetectorInfo.setPosition({i, j}, m_positions[i][j]);
+    }
+  }
+}
+
+void ScanningWorkspaceBuilder::buildRelativeRotationsForScans(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < outputDetectorInfo.size(); ++i) {
+    for (size_t j = 0; j < outputDetectorInfo.scanCount(i); ++j) {
+      auto position = outputDetectorInfo.position({i, j});
+      const auto rotation = Kernel::Quat(m_instrumentAngles[j], m_rotationAxis);
```

Comment:
Can you double check what this does? `Eigen::Quaternion` requires the axis to be normalized, otherwise you get garbage (a quaternion that is not representing a rotation), not sure about `Kernel::Quat`.